teps/0056-pipelines-in-pipelines.md

```diff
@@ -352,6 +363,451 @@ main-`PipelineRun` to be able to fetch and access the sub-`PipelineRun`'s full s
 cancellation.
 
+## Proposal
+
+This proposal focuses on enabling specification and execution of a `Pipeline` in a `Pipeline`. This section focuses 
+on the API change, see the [design](#design) section below for further details.
+
+### Specification
+
+To support defining `Pipelines` in `Pipelines` at authoring time, we propose adding `PipelineRef` and `PipelineSpec` 
+fields to [`PipelineTask`][pipelinetask] alongside `TaskRef` and `TaskSpec`. The `PipelineSpec` field will take a new 
+type, `EmbeddedPipeline`, similar to the [`EmbeddedTask`][embeddedtask] type, in order to support specifying metadata
+for child `PipelineRuns`.
+
+```go
+type EmbeddedPipeline struct {
+    PipelineSpec *PipelineSpec        `json:",inline,omitempty"`
+    Metadata     PipelineTaskMetadata `json:"metadata,omitempty"`
+}
+```
+
+```go
+type PipelineTask struct {
+    Name            string                `json:"name,omitempty"`
+    TaskRef         *TaskRef              `json:"taskRef,omitempty"`
+    TaskSpec        *EmbeddedTask         `json:"taskSpec,omitempty"`
+    PipelineRef     *PipelineRef          `json:"pipelineRef,omitempty"`
+    PipelineSpec    *EmbeddedPipeline     `json:"pipelineSpec,omitempty"`
+    ...
+}
+```
+
+For example, the [example](#fanning-out-pipelines) described above can be solved using a `Pipeline` named
+`security-scans` which is run within a `Pipeline` named `clone-scan-notify` where the `PipelineRef` is used.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: security-scans
+spec:
+  tasks:
+    - name: scorecards
+      taskRef:
+        name: scorecards
+    - name: codeql
+      taskRef:
+        name: codeql
+---
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: clone-scan-notify
+spec:
+  tasks:
+    - name: git-clone
+      taskRef:
+        name: git-clone
+    - name: security-scans
+      pipelineRef:
+        name: security-scans
+    - name: notification
+      taskRef:
+        name: notification
+```
+
+The above example can be modified to use `PipelineSpec` instead of `PipelineRef` if the user would prefer to embed the
+`Pipeline` specification:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: clone-scan-notify
+spec:
+  tasks:
+    - name: git-clone
+      taskRef:
+        name: git-clone
+    - name: security-scans
+      pipelineSpec:
+        tasks:
+          - name: scorecards
+            taskRef:
+              name: scorecards
+          - name: codeql
+            taskRef:
+              name: codeql
+    - name: notification
+      taskRef:
+        name: notification
+````
+
+To support declaring configuration used to create a `PipelineRun` from a `Pipeline` in a `Pipeline` at run time, we 
+propose creating a new `PipelineRunSpecs` field and adding it to [`PipelineRunSpec`][pipelinerunspec] field alongside 
+[`TaskRunSpecs`][taskrunspecs].
```

Comment:
we discussed this in the API WG this morning, can we avoid introducing a special `pipelineRunSpec` at the runtime?

Also, for the first phase of implementation, please feel free to support limited functionalities i.e. no support for timeouts.

Reply:
I tend to agree with @pritidesai. Timeouts is also an interesting example: what is the use case to have different timeouts for an inner pipeline ? What if the inner pipeline's timeout is bigger than the outer pipeline timeout ? I think we may not want to ask those questions initially, nor support those initialy.

Reply:
My perspective is that the sum of the timeouts of objects (Tasks, Runs, Pipelines) within a Pipeline should be less than or equal to that of the Pipeline -- so we'd add validation to check that we handle the Pipelines in Pipelines timeouts correctly

Though I agree that we can descope the runtime configuration from this TEP and explore it further in future iterations, what do you think @abayer?